### PR TITLE
Limit login attempts by a user

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## Description

Login login attempts by a user
## Impacts

User login into the captive portal
## NEWS file entries

New Features
++++++++++++
Allow limiting the user login attempts
## Delete branch after merge

NO

### CODE REVIEW

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm [hunk 1]
```diff
@@ -167,6 +167,7 @@ sub login : Local : Args(0) {
 
         # External authentication
         $c->forward('validateLogin');
+        $c->forward('testLoginRetryLimit');
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm]: Would be more enforceLoginRetryLimit since that method detaches if the user should be locked

**Comment by julsemaan** [comment]
Doesn't look like the field has been included in the portal template

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm [hunk 2]
```diff
@@ -178,6 +179,24 @@ sub login : Local : Args(0) {
...
+
+=cut
+
+sub testLoginRetryLimit : Private {
+    my ( $self, $c ) = @_;
+    my $username = $c->request->param("username");
+    if($username) {
+        if ($self->reached_retry_limit($c, "login_retries", $c->profile->{'_login_attempt_retry_limit'})) {
+            $c->log->info("Max tries reached login code for $username");
+            $c->stash(txt_validation_error => i18n_format($GUEST::ERRORS{$GUEST::ERROR_MAX_RETRIES}));
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm]: It doesn't show that message when you reach the limit.

It makes a silent failure.

Comment (jrouzierinverse) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm]: This has been fixed

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm (hunk 1 above)
Comment (jrouzierinverse) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm]: Done

**Comment by jrouzierinverse** [comment]
Please review

File: html/captive-portal/lib/captiveportal/PacketFence/Model/Portal/Session.pm [hunk 3]
```diff
@@ -110,8 +110,7 @@ sub ACCEPT_CONTEXT {
     my $options;
     my $mgmt_ip = $management_network->{'Tvip'} || $management_network->{'Tip'} if $management_network;
     my $destination_url = $request->param('destination_url');
-
-    if( defined ( my $last_uri = $r->pnotes('last_uri') )) {
+    if(  ref($r) ne 'GLOB' && ref($r) ne 'FileHandle' && defined ( my $last_uri = $r->pnotes('last_uri') )) {
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/Model/Portal/Session.pm]: Could you please explain this line with a comment.

Also what does this have to do with the last_uri ?

Comment (jrouzierinverse) [inline html/captive-portal/lib/captiveportal/PacketFence/Model/Portal/Session.pm]: Bad commit will update that

**Comment by extrafu** [comment]
Bump?

File: html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm [hunk 4]
```diff
@@ -254,6 +254,18 @@ has_field 'sms_pin_retry_limit' =>
              help => 'Maximum number of times a user can retry a SMS PIN before having to request another PIN. A value of 0 disables the limit.' },
 
   );
+=head2 login_attempt_retry_limit
+
+=cut
+
+has_field 'login_attempt_retry_limit' =>
```
Comment (cgx) [inline html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm]: How about simply `login_attempt_limit`?

**Comment by julsemaan** [comment]
please rebase the pull request

File: html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm [hunk 5]
```diff
@@ -254,6 +254,18 @@ has_field 'sms_pin_retry_limit' =>
              help => 'Maximum number of times a user can retry a SMS PIN before having to request another PIN. A value of 0 disables the limit.' },
 
   );
+=head2 login_attempt_retry_limit
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm]: There's no description for that sub
*variable

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm [hunk 6]
```diff
@@ -178,6 +179,24 @@ sub login : Local : Args(0) {
...
+=head2 enforceLoginRetryLimit
+
+Limit the amount of time a user can retry a password
+
+=cut
+
+sub enforceLoginRetryLimit : Private {
+    my ( $self, $c ) = @_;
+    my $username = $c->request->param("username");
+    if($username) {
```
Comment (cgx) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm]: Missing a space after the `if`.

File: html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm [hunk 7]
```diff
@@ -254,6 +254,18 @@ has_field 'sms_pin_retry_limit' =>
...
+
+=cut
+
+has_field 'login_attempt_retry_limit' =>
+  (
+    type => 'PosInteger',
+    label => 'Login Attempt Retry Limit',
+    default => 0,
+    tags => { after_element => \&help,
+             help => 'Maximum number of times a user attempt to login before being blocked. A value of 0 disables the limit.' },
```
Comment (cgx) [inline html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm]: Simplify this: `Limit the number of login attempts. A value of 0 disables the limit.`

**Comment by jrouzierinverse** [comment]
Please review
